Add step to drop plaintext columns

fimac opened the pull request:
- Adds step to drop plaintext fields
- Adds Ransack fix so filters continue to work
- Adds in Rails creds step

fimac committed:
add step to drop plaintext columns

fimac committed:
remove ignore columns workaround in model

fimac committed:
add ransack step to update admin file and model so filters continue to work

CDThomas commented on README.md:
````diff
@@ -188,20 +188,28 @@ Note down the client key somewhere safe, like a password vault.
 ```
 
-If you're not you can export the variables by running:
+Run:
````
What's the reason for removing the note about not using direnv here? The original text looks correct to me.

Alternatively, we could probably remove mentions of direnv in the demo and assume that the user knows how to set env vars. We could say "set these env vars" and let the user set them however they prefer.

fimac replied:
hmm, I think this may have been unintentional, as it doesn't read right. I was copying and pasting a bit over from the docs repo.

I think it's good to have it in there, trying to be helpful.

fimac replied:
I'll revert it back.

CDThomas commented on README.md:
```diff
@@ -402,6 +406,91 @@ Push this configration to CipherStash:
+
+  def self.full_name_filter_contains(value)
+
+    where("full_name LIKE ?", "#{sanitize_sql_like(value)}")
+  end
+
+  def self.full_name_filter_equals(value)
+
+    where(full_name: "#{sanitize_sql_like(value)}")
+  end
+
+  def self.email_filter_contains(value)
+
+    where("email LIKE ?", "#{sanitize_sql_like(value)}")
+  end
+
+  def self.email_filter_equals(value)
+
+    where(email: "#{sanitize_sql_like(value)}")
+  end
+
+  def self.allergies_filter_contains(value)
+
+    where("allergies LIKE ?", "#{sanitize_sql_like(value)}")
+  end
+
+  def self.allergies_filter_equals(value)
+
+    where(allergies: "#{sanitize_sql_like(value)}")
+  end
+
+  def self.medications_filter_contains(value)
+
+    where("medications LIKE ?", "#{sanitize_sql_like(value)}")
+  end
+
+  def self.medications_filter_equals(value)
+
+    where(medications: "#{sanitize_sql_like(value)}")
+  end
```
Nit: these methods probably don't need the extra whitespace.
```suggestion
  def self.full_name_filter_contains(value)
    where("full_name LIKE ?", "#{sanitize_sql_like(value)}")
  end

  def self.full_name_filter_equals(value)
    where(full_name: "#{sanitize_sql_like(value)}")
  end

  def self.email_filter_contains(value)
    where("email LIKE ?", "#{sanitize_sql_like(value)}")
  end

  def self.email_filter_equals(value)
    where(email: "#{sanitize_sql_like(value)}")
  end

  def self.allergies_filter_contains(value)
    where("allergies LIKE ?", "#{sanitize_sql_like(value)}")
  end

  def self.allergies_filter_equals(value)
    where(allergies: "#{sanitize_sql_like(value)}")
  end

  def self.medications_filter_contains(value)
    where("medications LIKE ?", "#{sanitize_sql_like(value)}")
  end

  def self.medications_filter_equals(value)
    where(medications: "#{sanitize_sql_like(value)}")
  end
```

CDThomas commented on README.md:
````diff
@@ -402,6 +406,91 @@ Push this configration to CipherStash:
+Add this to the `patients.rb` file:
+
+```ruby
+# app/admin/patients.rb
````
I'm starting to wonder if it'd be easier to check in some of these changes and include them from the starting point.

We could have some comments in the source files that explain some of the active admin details for those that are curious. That way we can highlight the important parts that are most relevant to the demo in the README and leave out some of the active admin tedium.

auxesis replied:
@CDThomas that's a really good point. Way easier to uncomment than copy-paste.

fimac replied:
@CDThomas @auxesis I thought of doing that, but if they try and run the string filters in `plaintext-duplicate` mode, they will not work. Because the custom functions we add, to keep the string filters working in encrypted-duplicate and encrypted mode, strip out the`%` from the LIKE queries.

fimac replied:
Oh, you mean just add it, but commented out?

fimac replied:
I'll add it into the demo, commented out with notes.

I'll need to add something in the README around it though, to point them to uncomment it.

CDThomas replied:
> Oh, you mean just add it, but commented out?

Ah, I meant have any active-admin related code included in the project from the start (not commented out), but I didn't think about the `LIKE` query issues.

Commenting it out might work though.

I'm really tempted to get `LIKE` queries working for both `plaintext-duplicate` and `enrypted*` modes without editing files. If we don't want to deal with making the actual `LIKE` implementation more robust yet, I wonder if it'd be straightforward to handle it in the match indexer instead. We could add support for `char_filters` and strip out `%`s and `_`s or something.

CDThomas replied:
> We could add support for char_filters and strip out `%`s and `_`s or something.

Since you added support for `ILIKE` that might mean that we wouldn't even need the custom active admin filter methods, too @fimac.

fimac replied:
We still need them, because the filters send through the ILIKE query with the `%`, so the match query won't work. So we need to intercept the query anyway, so we're sending through just the value. Unless i'm missing something with how the match query is implemented?

fimac replied:
> I'm really tempted to get `LIKE` queries working for both `plaintext-duplicate` and `enrypted*` modes without editing files. If we don't want to deal with making the actual `LIKE` implementation more robust yet, I wonder if it'd be straightforward to handle it in the match indexer instead. We could add support for `char_filters` and strip out `%`s and `_`s or something.

This would be ideal, well until we can do the actual LIKE/ILIKE implementation, I can create a card if there isn't one already?

CDThomas replied:
`char_filter` support seems like something we could want in the long term anyway, so seems reasonable to me. It feels less hacky than stripping out lead/trailing `%`s in the driver code.

Maybe run it by James and/or Dan too and card it up if it sounds good to them. I don't think it'd be too hard to implement.

CDThomas commented on README.md:
```diff
@@ -467,17 +556,99 @@ Use the filters on the side to perform queries.
+>
+> **Once you remove the plaintext columns, anything that hasn't been encrypted will be lost.**
+>
+> Before you run the remove column step, it is very important that you:
```
Nit: phrasing
```suggestion
> Before you drop plaintext columns in a real-world application, it is very important that you:
```

fimac committed:
revert change to line to before running env vars

fimac committed:
remove extra whitespace

fimac committed:
phrasing

Co-authored-by: Drew Thomas <7672425+CDThomas@users.noreply.github.com>

fimac committed:
refer to the product consistently

fimac committed:
add timeline to set expectations

fimac committed:
link to active admin

fimac committed:
explain what the mode does

fimac committed:
make the statement clearer

fimac committed:
add note about uncommenting types in patient model

fimac committed:
add in code to uncomment in patients and patient files

fimac committed:
update copy around adding to uncommenting code

fimac commented:
@CDThomas @auxesis This is ready for a re review, I have:
- updated this with all the suggestions @auxesis made in the docs PR.
- addressed the suggestions from @CDThomas  in this PR.
- I've added in the code to the patients.rb file and the patient model, so they only have to be uncommented. This is a bit clunky, but should only be for the short term. I'll create a card to look into the LIKE/ILIKE implementations, so we don't have to do any code changes.

auxesis reviewed (approved):
Big fan of this, thanks for putting it together!
